server/charm/src/charm.py

```diff
@@ -3,62 +3,80 @@
 # See LICENSE file for licensing details.
 #
 # Learn more at: https://juju.is/docs/sdk
-
-"""Charm the service.
-
-Refer to the following tutorial that will help you
-develop a new k8s charm using the Operator Framework:
-
-https://juju.is/docs/sdk/create-a-minimal-kubernetes-charm
+"""
+Hardware API Charm
 """
 
 import logging
 
 import ops
+from charms.nginx_ingress_integrator.v0.nginx_route import require_nginx_route
 
 # Log messages can be retrieved using juju debug-log
 logger = logging.getLogger(__name__)
 
 VALID_LOG_LEVELS = ["info", "debug", "warning", "error", "critical"]
 
 
-class CharmCharm(ops.CharmBase):
+class HardwareApiCharm(ops.CharmBase):
     """Charm the service."""
 
     def __init__(self, *args):
         super().__init__(*args)
-        self.framework.observe(self.on['hardware_api'].pebble_ready, self._on_hardware_api_pebble_ready)
+        self._setup_nginx()
+        self.framework.observe(
+            self.on["hardware-api"].pebble_ready,
+            self._on_hardware_api_pebble_ready,
+        )
         self.framework.observe(self.on.config_changed, self._on_config_changed)
 
+    def _setup_nginx(self):
+        require_nginx_route(
+            charm=self,
+            service_hostname=self.config["hostname"],
+            service_name=self.app.name,
+            service_port=int(self.config["port"]),
+        )
+
     def _on_hardware_api_pebble_ready(self, event: ops.PebbleReadyEvent):
         container = event.workload
-        container.add_layer("hardware_api", self._pebble_layer, combine=True)
+        container.add_layer("hardware-api", self._pebble_layer, combine=True)
         container.replan()
         self.unit.status = ops.ActiveStatus()
 
     def _on_config_changed(self, event: ops.ConfigChangedEvent):
         log_level = self.model.config["log-level"].lower()
 
         if log_level in VALID_LOG_LEVELS:
-            container = self.unit.get_container("hardware_api")
+            container = self.unit.get_container("hardware-api")
             if container.can_connect():
-                container.add_layer("hardware_api", self._pebble_layer, combine=True)
+                container.add_layer(
+                    "hardware-api", self._pebble_layer, combine=True
+                )
                 container.replan()
                 logger.debug("Log level changed to '%s'", log_level)
                 self.unit.status = ops.ActiveStatus()
             else:
                 event.defer()
                 self.unit.status = ops.WaitingStatus("waiting for Pebble API")
         else:
-            self.unit.status = ops.BlockedStatus("invalid log level: '{log_level}'")
+            self.unit.status = ops.BlockedStatus(
+                "invalid log level: '{log_level}'"
+            )
+
+    @property
+    def _app_environment(self):
+        """Environment variables needed by the application."""
+        env = {}
+        return env
```

Comment:
are we using this anywhere? or is it that we will be using it soon

Reply:
It was already trying to use _app_environment but the property for it didn't exist yet. I thought about ripping it out, but I suspect we might want things like sentry at some point, so I thought it might be better to just add a stub for it for now.